[OGR] reset layername after SELECT filter is removed

elpaso opened the pull request:
Fix #56345

When a provider subset string like "SELECT* FROM table WHERE ...." was removed, no `layername=xxxx` was added to the URI, making it ambiguous in case of multiple tables in a dataset (e.g. GPKG).

elpaso committed:
[OGR] reset layername after SELECT filter is removed

Fix #56345

rouault commented on src/core/vector/qgsvectorlayer.cpp:
```diff
@@ -1266,6 +1266,7 @@ bool QgsVectorLayer::setSubsetString( const QString &subset )
 
   if ( res )
   {
+    mWkbType = mDataProvider->wkbType();
```
that change seems unrelated

rouault replied:
or maybe it is. please ignore

nirvn commented:
@elpaso , not a bad solution at all :)

elpaso committed:
Don't fallback when no layername was set in the first place

Mainly because it breaks a lot of tests, it would probably
be a good idea to revert this commit eventually.

nyalldawson commented on src/core/vector/qgsvectorlayer.cpp:
```diff
@@ -1266,6 +1266,7 @@ bool QgsVectorLayer::setSubsetString( const QString &subset )
 
   if ( res )
   {
+    mWkbType = mDataProvider->wkbType();
```
I think there's a little more required here to support changing geometry types via queries -- eg we'd need to rebuild the layer renderer from a default if the geometry type changes. Could this change be split off into a separate PR?

elpaso replied:
done

elpaso committed:
Update test after fallback removal

elpaso committed:
Remove geometry type change

nyalldawson commented:
@elpaso looks like a valid test failure here.

elpaso committed:
Add test on feature geometry type

elpaso commented:
> @elpaso looks like a valid test failure here.

I cannot reproduce the test failure locally.

agiudiceandrea commented:
@elpaso why don't just add the `layername=` part when the layer is added via Browser or DB manager? Should'n it be easier to implement?

elpaso commented:
> @elpaso why don't just add the `layername=` part when the layer is added via Browser or DB manager? Should'n it be easier to implement?

That would not fix the issue because in case of a "SELECT..." filter a different table could be selected and in that case we want to set the correct "layername"  when the filter is removed.

This patch fixes the issue by parsing the query filter and getting the table from there. It will work in most cases, it won't in case of multiple tables in the FROM clause.

Anyway, I cannot reproduce the CI test error locally and I don't know when I'll find the time to work on this issue again.

agiudiceandrea commented:
My suggestion is to be consistent with the the already present behaviour:
- add a GPKG sub layer to the project -> source: `/path/file_name.gpkg|layername=sub_layer_name`
- add any SQL filter to the layer (even selecting from another sub layer) -> source: `/path/file_name.gpkg|layername=sub_layer_name|subset=the_SQL_query`
- remove the filter from the layer -> source: `/path/file_name.gpkg|layername=sub_layer_name`

Otherwise I think also the already present behaviour should be changed.
